capstone/src/main/java/com/google/step/clients/MetricsClient.java

```diff
@@ -0,0 +1,288 @@
+            try {
+                Entity entity = new Entity("PageViews");
+                entity.setProperty("restaurantKey", restaurantKey);
+                entity.setProperty("restaurantName", restaurantName);
+                entity.setProperty("year", year);
+                entity.setProperty("week", week);
+                entity.setProperty("count", 0);
+                dataStore.put(transaction, entity);
+
+                // Set the count for the return map as 0, since the entity doesn't exist.
+                resultMap.put("count", 0);
+                transaction.commit();
+            } finally {
+                if (transaction.isActive()) {
+                    transaction.rollback();
+                }
+            }
+        } else {
+            // If the results didn't come back null, give the count.
+            resultMap.put("count", results.getProperty("count"));
+        }
+
+        resultMap.put("year", year);
+        resultMap.put("week", week);
+
+        return resultMap;
+    }
+
+    /**
+     * Gets the weekly pageViews for the entire year of a specific restaurant.
+     * @param year we are looking for
+     * @param restaurantKey of restaurant we want the data.
+     * @return list of maps in the following way: List<Map<String, Object>>
+     *          [
+     *                 {
+     *                      "year": <String, the year number>,
+     *                      "week": <String, this week's number in relation to the year.>,
+     *                      "count": <String, amount of page views the restaurant had.>
+     *                  },
+     *                  ...,
+     *          ]
+     */
+    public List<Map<String, Object>> getYearRestaurantPageViews(int year, String restaurantKey) {
+
+        CompositeFilter compositeFilter = new CompositeFilter(Query.CompositeFilterOperator.AND,
+                Arrays.asList(
+                        new FilterPredicate("restaurantKey", Query.FilterOperator.EQUAL, restaurantKey),
+                        new FilterPredicate("year", Query.FilterOperator.EQUAL, year)
+                ));
+        Query query = new Query("PageViews").setFilter(compositeFilter).addSort("week", Query.SortDirection.ASCENDING);
+        PreparedQuery preparedQuery = dataStore.prepare(query);
+        List<Entity> results = preparedQuery.asList(FetchOptions.Builder.withDefaults());
+
+        List<Map<String, Object>> resultList = new ArrayList<>();
+
+        Map<String, Object> map;
+        for (Entity entity : results) {
+            map = new HashMap<>();
+            map.put("week", entity.getProperty("week").toString());
+            map.put("count", Integer.parseInt(entity.getProperty("count").toString()));
+            resultList.add(map);
+        }
+
+        return resultList;
+    }
+
+    /**
+     * Gets all pageViews for the entire system.
+     * @return returns a Map<String, Object> in the following format:
+     *          {
+     *              "firstDate":
```

Comment:
Designing APIs is a tough racket. You're building an interface that various customers will be using. You need to have robust constructs that are flexible and generalize the concepts, BUT that also define some rules, and remain as simple as possible.

When owning and maintaining an API, customers will often ask you to add this and that to it, because it would be a bit convenient for them. However, in this case, if I owned the "MetricsClient" API and a customer asked me to include a "firstDate", I would ask some questions about why it's needed.

Why is this valuable?

What does this add that reading the first entry in the clickData list doesn't add?

Why is the week BEFORE first date any different than "the week in the middle of the data where no clicks happened so there's no result"?

Basically: The information that "firstDate" gives you is trivially easy to retrieve from the clickData that you're already returning (which should be sorted, as we've previously stated, and as you've previously done in the previous API method). And, it kind of complicates the API, because any future prospective customer would ask, "why is that there? is it needed? is the response data more complex than I imagine?" It will add confusion and very little value.

Would things not work if you just removed this entirely?

Reply:
I'll add a drive-by comment here because I asked Martin to add firstDate :) I had asked for it because what I am planning to do with the visualization will require me to know the first date we have data before looping through and further processing all the clickData. Thus I thought that rather than looping through to find the minimum date once, doing what I need to do with it, then having to loop through all the data again once things are set up, I'd ask Martin to send it back to me since it'd be relatively easy to find/keep track of while he was building the response.
However, like you said, I can certainly retrieve it from the click data in the visualization code where I actually need it, rather than have it here in the API. It's great to learn that designing an API shouldn't necessarily be about specific things a customer (or teammate) wants!

Reply:
This has been taken out, but the sorting within each list still exists, that way it's not too hard to find the earliest date from the frontend.

Reply:
Great work, team! :D